SimPEG/ObjectiveFunction.py

```diff
@@ -0,0 +1,258 @@
+    def _test_deriv2(self, x=None, num=4, plotIt=False, **kwargs):
+        print('Testing {0!s} Deriv2'.format(self.__class__.__name__))
+        if x is None:
+            if self.nP == '*':
+                x = np.random.randn(np.random.randint(1e2, high=1e3))
+            else:
+                x = np.random.randn(self.nP)
+
+        return checkDerivative(
+            lambda m: [self.deriv(m), self.deriv2(m)], x, num=num,
+            plotIt=plotIt
+        )
+
+    def test(self, x=None, num=4, plotIt=False, **kwargs):
+        deriv = self._test_deriv(x=x, num=num, **kwargs)
+        deriv2 = self._test_deriv2(x=x, num=num, plotIt=False, **kwargs)
+        return (deriv & deriv2)
+
+    def __add__(self, objfct2):
+
+        if isinstance(objfct2, Utils.Zero):
+            return self
+
+        if not isinstance(objfct2, BaseObjectiveFunction):
+            raise Exception(
+                "Cannot add type {} to an objective function. Only "
+                "ObjectiveFunctions can be added together".format(
+                    objfct2.__class__.__name__
+                )
+            )
+
+        if isinstance(self, ComboObjectiveFunction):
+
+            if isinstance(objfct2, ComboObjectiveFunction):
+                objfctlist = self.objfcts + objfct2.objfcts
+                multipliers = self.multipliers + objfct2.multipliers
+
+            elif isinstance(objfct2, ObjectiveFunction):
+                objfctlist = self.objfcts.append(objfct2)
+                multipliers = self.multipliers.append(1)
+
+        else:
+            if isinstance(objfct2, ComboObjectiveFunction):
+                objfctlist = [self] + objfct2.objfcts
+                multipliers = [1] + objfct2.multipliers
+
+            else:
+                objfctlist = [self, objfct2]
+                multipliers = None
+
+        return ComboObjectiveFunction(
+            objfcts=objfctlist, multipliers=multipliers
+        )
+
+    def __radd__(self, objfct2):
+        return self+objfct2
+
+    def __mul__(self, multiplier):
+        return ComboObjectiveFunction([self], [multiplier])
+
+    def __rmul__(self, multiplier):
+        return self * multiplier
+
+    def __div__(self, denominator):
+        return self.__mul__(1./denominator)
+
+
+class ComboObjectiveFunction(BaseObjectiveFunction):
+
+    _multiplier_types = (float, None, Utils.Zero) + integer_types # Directive
+
+    def __init__(self, objfcts=[], multipliers=None, **kwargs):
+
+        self.objfcts = []
+        nP = '*'
+
+        for fct in objfcts:
+            assert isinstance(fct, BaseObjectiveFunction), (
+                "Unrecognized objective function type {} in objfcts. All "
+                "entries in objfcts must inherit from "
+                "ObjectiveFunction".format(fct.__class__.__name__)
+            )
+
+            # ensure all objective functions have the same nP
+            if fct.nP != '*':
+                if nP != '*':
+                    assert nP == fct.nP, (
```

Comment:
- the objective function does need to keep track of `nP`. If we are adding two objective functions together, they need to take the same size of input. 
- currently, the base objective function does not have a mapping. Currently only the `Regularization` has a mapping, but for joint inversions, a mapping might be necessary. 

So prior to merging, 
- [ ] the BaseObjective function should have a mapping